demos/GradientDescent.cpp

```diff
@@ -0,0 +1,138 @@
+//--------------------------------------------------------------------*- C++ -*-
+// clad - The C++ Clang-based Automatic Differentiator
+//
+// Demo of a simple gradient decent algorithm to predict the feautres for a
+// liner relationship. Clad is used to calculate the gradient of the
+// cost function.
+//
+//----------------------------------------------------------------------------//
+
+// To run the demo please type:
+// path/to/clang  -Xclang -add-plugin -Xclang clad -Xclang -load -Xclang \
+// path/to/libclad.so  -I../include/ -x c++ -lstdc++ -lm GradientDescent.cpp
+//
+// A typical invocation would be:
+// ../../../../obj/Debug+Asserts/bin/clang  -Xclang -add-plugin -Xclang clad \
+// -Xclang -load -Xclang ../../../../obj/Debug+Asserts/lib/libclad.dylib     \
+// -I../include/ -x c++ -lstdc++ -lm GradientDescent.cpp
+//
+// To plot the results install gnuplot and type:
+// gnuplot -e "plot 'dataset_gd.dat' with points pt 7; replot 'out_gd.dat' \
+// using 1:2 with lines; pause -1"
+
+#include <algorithm>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+// Necessary for clad to work include
+#include "clad/Differentiator/Differentiator.h"
+
+// The hypothesis function for the quadratic problem
+// theta_0, theta_1 are the parameters to learn and x is the input data
+double f(double theta_0, double theta_1, double x) {
+  return theta_0 + theta_1 * x;
+}
+
+// Structure to represent input dataset
+// x, y represents the input and output data respectively
+// size represents the number of samples to be generated
+struct Dataset {
+  std::vector<double> x;
+  std::vector<double> y;
+  size_t size = 1000;
+  double learning_rate = 1e-3;
+
+  // Populate the data set with random data
+  Dataset() {
+    // For plotting data later
+    std::ofstream out("dataset_gd.dat");
+
+    // For storing the randomized values
+    double rand_x, rand_y, t0, t1;
+
+    for (size_t i = 0; i < size; i++) {
+      rand_x = 3 * (float)(std::rand() % 100) / 100;
+      t0 = 9 + (float)(std::rand() % 100) / 100;
+      t1 = 2;
+      rand_y = f(t0, t1, rand_x);
+
+      x.push_back(rand_x);
+      y.push_back(rand_y);
+      out << x[i] << "\t" << y[i] << std::endl;
+    }
+
+    out.close();
+  }
+};
+
+// Function to perform a minimization step
+// theta_x are the hypothesis parameters, t is the generated dataset and
+// clad_grad is the gradient function generated by Clad
+template <typename T>
+void performStep(double &theta_0, double &theta_1, Dataset dt, T clad_grad) {
+  double J_theta[4], result[2] = {0, 0};
+
+  for (size_t i = 0; i < dt.size; i++) {
+    clad_grad.execute(theta_0, theta_1, dt.x[i], dt.y[i], J_theta);
```

Comment:
Possible fix for the `-O2` error is initializing `J_theta`elements to `0`.
```suggestion
    J_theta[0] = J_theta[1] = J_theta[2] = J_theta[3] = 0;
    clad_grad.execute(theta_0, theta_1, dt.x[i], dt.y[i], &J_theta[0], &J_theta[1], &J_theta[2], &J_theta[3]);
```

Reply:
@sudo-panda Thanks! That + tuning the hyperparameters solved it! Super silly issue :P